File: src/main/java/se/kth/spork/base3dm/ChangeSet.java
package se.kth.spork.base3dm;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Represents a change set in 3DM merge. While a change set in pure 3DM merge is just all content tuples and PCS
 * triples put together, this change set separates content and structure into separate sets, and also introduces
 * some helper functionality to keep track of conflicts and enable faster lookup of PCS triples.
 *
 * @author Simon Larsén
 */
public class ChangeSet<T extends ListNode,V> {
    private Map<T, Set<Pcs<T>>> successors;
    private Map<T, Set<Pcs<T>>> predecessors;
    private Map<T, Set<Content<T,V>>> content;
    private Map<T, T> classRepMap;
    private Set<Pcs<T>> pcsSet;
    private Map<Pcs<T>, Set<Pcs<T>>> structuralConflicts;

    @SuppressWarnings("unchecked")
    private final Set<Pcs<T>> EMPTY_PCS_SET = Collections.EMPTY_SET;

    @SuppressWarnings("unchecked")
    private final Set<Content<T,V>> EMPTY_CONTENT_SET = Collections.EMPTY_SET;

    /**
     * Create a T* from the provided trees, using the class representatives map to map each node to its class
     * representative.
     *
     * @param classRepMap A map mapping each node to its class representative.
     * @param getContent A function for getting content from
     * @param trees The trees to add to this T*.
     */
    @SafeVarargs
    public ChangeSet(Map<T, T> classRepMap, Function<T, V> getContent, Set<Pcs<T>>... trees) {
        this.classRepMap = classRepMap;
        successors = new HashMap<>();
        predecessors = new HashMap<>();
        content = new HashMap<>();
        pcsSet = new HashSet<>();
        Arrays.stream(trees).forEach(t -> add(t, getContent));
        structuralConflicts = new HashMap<>();
    }

    /**
     * @return The current state of the structure of the merge.
     */
    public Set<Pcs<T>> getPcsSet() {
        return new HashSet<>(pcsSet);
    }

    /**
     * @return The current state of the merge's node contents.
     */
    public Map<T, Set<Content<T,V>>> getContents() {
        return new HashMap<>(content);
    }

    /**
     * @param pcs A PCS triple.
     * @return All PCSes that are root conflicting with the provided PCS.
     */
    public List<Pcs<T>> getOtherRoots(Pcs<T> pcs) {
        return Stream.of(pcs.getPredecessor(), pcs.getSuccessor()).flatMap(
                node -> Stream.of(predecessors.get(node), successors.get(node))
                        .filter(Objects::nonNull)
                        .flatMap(Set::stream)
        ).filter(p -> !p.getRoot().equals(pcs.getRoot())).collect(Collectors.toList());
    }

    /**
     * @param pcs A PCS triple.
     * @return All PCSes that are successor conflicting with the provided PCS.
     */
    public List<Pcs<T>> getOtherSuccessors(Pcs<T> pcs) {
        return predecessors.getOrDefault(pcs.getPredecessor(), EMPTY_PCS_SET).stream()
                .filter(p -> !p.getSuccessor().equals(pcs.getSuccessor()))
                .collect(Collectors.toList());
    }

    /**
     * @param pcs A PCS triple.
     * @return All PCSes that are predecessor conflicting with the provided PCS.
     */
    public List<Pcs<T>> getOtherPredecessors(Pcs<T> pcs) {
        return successors.getOrDefault(pcs.getSuccessor(), EMPTY_PCS_SET).stream()
                .filter(p -> !p.getPredecessor().equals(pcs.getPredecessor()))
                .collect(Collectors.toList());
    }

    /**
     * @param node A node..
     * @return The content associated with the argument node, or an empty set if no content was associated.
     */
    public Set<Content<T,V>> getContent(T node) {
        return Collections.unmodifiableSet(content.getOrDefault(node, EMPTY_CONTENT_SET));
    }

    /**
     * @param pcs A PCS triple.
     * @return true iff the argument is contained in this T*.
     */
    public boolean contains(Pcs<T> pcs) {
        Set<Pcs<T>> matches = predecessors.get(pcs.getPredecessor());
        return matches != null && matches.contains(pcs);
    }

    /**
     * @param cont A Content container.
     * @return true iff the argument is contained in this T*.
     */
    public boolean contains(Content<T,V> cont) {
        return content.getOrDefault(cont.getContext().getPredecessor(), new HashSet<>()).contains(cont);
    }

    /**
     * Set the content for some pcs triple, overwriting anything that was there previously.
     *
     * @param node A node to associate the content with. This is the key in the backing map.
     * @param nodeContents A set of content values to associate with the node.
     */
    public void setContent(T node, Set<Content<T, V>> nodeContents) {
        content.put(node, nodeContents);
    }

    /**
     * Remove the PCS from all lookup tables, except for the contents table. Also remove it from the *-set.
     *
     * @param pcs A PCS triple.
     */
    public void remove(Pcs<T> pcs) {
        T pred = pcs.getPredecessor();
        T succ = pcs.getSuccessor();

        predecessors.get(pred).remove(pcs);
        successors.get(succ).remove(pcs);

        pcsSet.remove(pcs);
    }

    /**
     * @param cont Content to remove from this T*.
     */
    public void remove(Content<T,V> cont) {
        content.get(cont.getContext().getPredecessor()).remove(cont);
    }

    /**
     * Register a conflict in a bi-directional lookup table.
     *
     * @param first A PCS triple that conflicts with second.
     * @param second A PCS triple that conflicts with first.
     */
    public void registerStructuralConflict(Pcs<T> first, Pcs<T> second) {
        addToLookupTable(first, second, structuralConflicts);
        addToLookupTable(second, first, structuralConflicts);
    }

    /**
     * @return A copy of the internal structural conflicts lookup table.
     */
    public Map<Pcs<T>, Set<Pcs<T>>> getStructuralConflicts() {
        return new HashMap<>(structuralConflicts);
    }

    /**
     * Check if a PCS is involved in a structural conflict.
     *
     * @param pcs A PCS triple.
     * @return true iff the argument is involved in a structural conflict.
     */
    public boolean inStructuralConflict(Pcs<T> pcs) {
        return structuralConflicts.containsKey(pcs);
    }

    /**
     * Add a tree to this T*. This entails converting the entire PCS tree to its class representatives. Each
     * "class representative PCS" is then added to the predecessor and successor lookup tables, and their contents
     * are added to the content lookup table.
     *
     * @param tree A PCS tree structure.
     * @param getContent A function that returns the content of a T node.
     */
    private void add(Set<Pcs<T>> tree, Function<T, V> getContent) {
        for (Pcs<T> pcs : tree) {
            Pcs<T> classRepPcs = addToStar(pcs);
            T pred = pcs.getPredecessor();
            T classRepPred = classRepPcs.getPredecessor();
            T classRepSucc = classRepPcs.getSuccessor();

            if (classRepPred != null || classRepSucc != null) { // don't map leaf nodes
                addToLookupTable(classRepPred, classRepPcs, predecessors);
                addToLookupTable(classRepSucc, classRepPcs, successors);
            }
            if (!pred.isListEdge()) {
                Content<T,V> c = new Content<T,V>(pcs, getContent.apply(pred));
                addToLookupTable(classRepPred, c, content);
            }
        }
    }

    private Pcs<T> addToStar(Pcs<T> pcs) {
        T root = pcs.getRoot();
        T pred = pcs.getPredecessor();
        T succ = pcs.getSuccessor();
        Pcs<T> classRepPcs = new Pcs<T>(classRepMap.get(root), classRepMap.get(pred), classRepMap.get(succ), pcs.getRevision());
        pcsSet.add(classRepPcs);
        return classRepPcs;
    }

    private static <K, V> void addToLookupTable(K key, V val, Map<K, Set<V>> lookup) {
        Set<V> values = lookup.getOrDefault(key, new HashSet<V>());
        if (values.isEmpty())
            lookup.put(key, values);
        values.add(val);
    }
}

File: src/main/kotlin/se/kth/spork/base3dm/ChangeSet.kt
package se.kth.spork.base3dm

import java.util.Collections

/**
 * Represents a change set in 3DM merge. While a change set in pure 3DM merge is just all content tuples and PCS
 * triples put together, this change set separates content and structure into separate sets, and also introduces
 * some helper functionality to keep track of conflicts and enable faster lookup of PCS triples.
 *
 * @author Simon Larsén
 */
class ChangeSet<T : ListNode, V>(private val classRepMap: Map<T, T>, getContent: (T) -> V, vararg trees: Set<Pcs<T>>) {
    private val successors: MutableMap<T, MutableSet<Pcs<T>>>
    private val predecessors: MutableMap<T, MutableSet<Pcs<T>>>

    private val _structuralConflicts: MutableMap<Pcs<T>, MutableSet<Pcs<T>>>
    val structuralConflicts: Map<Pcs<T>, Set<Pcs<T>>>
        get() = _structuralConflicts.toMap()

    private val _pcsSet: MutableSet<Pcs<T>>
    val pcsSet: Set<Pcs<T>>
        get() = _pcsSet.toSet()

    private val _content: MutableMap<T, MutableSet<Content<T, V>>>
    val contents: Map<T, MutableSet<Content<T, V>>>
        get() = HashMap(_content)

    /**
     * @param pcs A PCS triple.
     * @return All PCSes that are root conflicting with the provided PCS.
     */
    fun getOtherRoots(pcs: Pcs<T>): List<Pcs<T>> {
        return sequenceOf(pcs.predecessor, pcs.successor).flatMap { node ->
            sequenceOf(predecessors[node], successors[node]).filterNotNull().flatMap {
                it.asSequence()
            }.filter {
                it.root != pcs.root
            }
        }.toList()
    }

    /**
     * @param pcs A PCS triple.
     * @return All PCSes that are successor conflicting with the provided PCS.
     */
    fun getOtherSuccessors(pcs: Pcs<T>): List<Pcs<T>> {
        return (predecessors[pcs.predecessor] ?: setOf<Pcs<T>>()).filter {
            it.successor != pcs.successor
        }.toList()
    }

    /**
     * @param pcs A PCS triple.
     * @return All PCSes that are predecessor conflicting with the provided PCS.
     */
    fun getOtherPredecessors(pcs: Pcs<T>): List<Pcs<T>> {
        return (successors[pcs.successor] ?: emptySet<Pcs<T>>()).filter {
            it.predecessor != pcs.predecessor
        }.toList()
    }

    /**
     * @param node A node..
     * @return The content associated with the argument node, or an empty set if no content was associated.
     */
    fun getContent(node: T): Set<Content<T, V>> {
        return Collections.unmodifiableSet(_content[node] ?: emptySet())
    }

    /**
     * @param pcs A PCS triple.
     * @return true iff the argument is contained in this T*.
     */
    operator fun contains(pcs: Pcs<T>): Boolean {
        return predecessors[pcs.predecessor]?.contains(pcs) == true
    }

    /**
     * @param cont A Content container.
     * @return true iff the argument is contained in this T*.
     */
    operator fun contains(cont: Content<T, V>): Boolean {
        return _content[cont.context.predecessor]?.contains(cont) == true
    }

    /**
     * Set the content for some pcs triple, overwriting anything that was there previously.
     *
     * @param node A node to associate the content with. This is the key in the backing map.
     * @param nodeContents A set of content values to associate with the node.
     */
    fun setContent(node: T, nodeContents: MutableSet<Content<T, V>>) {
        _content[node] = nodeContents
    }

    /**
     * Remove the PCS from all lookup tables, except for the contents table. Also remove it from the *-set.
     *
     * @param pcs A PCS triple.
     */
    fun remove(pcs: Pcs<T>) {
        val pred = pcs.predecessor
        val succ = pcs.successor
        predecessors[pred]!!.remove(pcs)
        successors[succ]!!.remove(pcs)
        _pcsSet.remove(pcs)
    }

    /**
     * @param cont Content to remove from this T*.
     */
    fun remove(cont: Content<T, V>) {
        _content[cont.context.predecessor]!!.remove(cont)
    }

    /**
     * Register a conflict in a bi-directional lookup table.
     *
     * @param first A PCS triple that conflicts with second.
     * @param second A PCS triple that conflicts with first.
     */
    fun registerStructuralConflict(first: Pcs<T>, second: Pcs<T>) {
        addToLookupTable(first, second, _structuralConflicts)
        addToLookupTable(second, first, _structuralConflicts)
    }

    /**
     * Add a tree to this T*. This entails converting the entire PCS tree to its class representatives. Each
     * "class representative PCS" is then added to the predecessor and successor lookup tables, and their contents
     * are added to the content lookup table.
     *
     * @param tree A PCS tree structure.
     * @param getContent A function that returns the content of a T node.
     */
    private fun add(tree: Set<Pcs<T>>, getContent: (T) -> V) {
        for (pcs in tree) {
            val classRepPcs = addToStar(pcs)
            val pred = pcs.predecessor
            val classRepPred = classRepPcs.predecessor
            val classRepSucc = classRepPcs.successor
            addToLookupTable(classRepPred, classRepPcs, predecessors)
            addToLookupTable(classRepSucc, classRepPcs, successors)
            if (!pred.isVirtual) {
                val c = Content(pcs, getContent(pred), pred.revision)
                addToLookupTable(classRepPred, c, _content)
            }
        }
    }

    private fun addToStar(pcs: Pcs<T>): Pcs<T> {
        val root = classRepMap[pcs.root] ?: error("${pcs.root} not in class representatives")
        val pred = classRepMap[pcs.predecessor] ?: error("${pcs.predecessor} not in class representatives")
        val succ = classRepMap[pcs.successor] ?: error("${pcs.successor} not in class representatives")
        val classRepPcs = Pcs(root, pred, succ, pcs.revision)
        _pcsSet.add(classRepPcs)
        return classRepPcs
    }

    companion object {
        private fun <K, V> addToLookupTable(key: K, `val`: V, lookup: MutableMap<K, MutableSet<V>>) {
            val values = lookup.getOrDefault(key, HashSet())!!
            if (values.isEmpty()) lookup[key] = values
            values.add(`val`)
        }
    }

    /**
     * Create a T* from the provided trees, using the class representatives map to map each node to its class
     * representative.
     *
     * @param classRepMap A map mapping each node to its class representative.
     * @param getContent A function for getting content from
     * @param trees The trees to add to this T*.
     */
    init {
        successors = HashMap()
        predecessors = HashMap()
        _content = HashMap()
        _pcsSet = HashSet()
        _structuralConflicts = HashMap()
        trees.forEach { t: Set<Pcs<T>> -> add(t, getContent) }
    }
}
